### PR TITLE
Fix logic around State/Province Zip/Postal requirements

### PR DESCRIPTION
**Pull request by markolson**
Once this is merged in, the Address form sections for both the
vet and their spouse will have differing requirements based on
if the country selected in USA. If so, State and ZIP will be
required. If USA, Canada or Mexico are selected, State/Province
will be a dropdown of permitted values, otherwise it will become
a free text entry field, and ZIP/Postal will be optional.

Addresses are also included in Spouse Associations now.

### CODE REVIEW

File: src/client/components/questions/Address.jsx [hunk 1]
```diff
@@ -83,23 +94,56 @@ class Address extends React.Component {
...
+    if (this.state.requireStateZip) {
+      secondaryForm = (
+        <div>
+          <ErrorableSelect errorMessage={this.validateAddressField(this.props.value.state) ? undefined : 'Please enter a valid state'}
+              label="State/Province"
+              name="state"
+              autocomplete="address-level1"
+              options={stateList}
+              value={this.props.value.state}
+              required={this.state.requireStateZip ? this.props.required : false}
```
Comment (austin-martinez) [inline src/client/components/questions/Address.jsx]: this conditional will always be true as this is in a block dependent on the condition being true L101

File: src/client/components/questions/Address.jsx [hunk 2]
```diff
@@ -83,23 +94,56 @@ class Address extends React.Component {
...
 
-        <ErrorableTextInput errorMessage={this.validateAddressField(this.props.value.zipcode) ? undefined : 'Please enter a valid ZIP code'}
-            label="ZIP Code"
-            name="zip"
-            autocomplete="postal-code"
-            field={this.props.value.zipcode}
-            required={this.props.required}
-            onValueChange={(update) => {this.handleChange('zipcode', update);}}/>
+          <ErrorableTextInput errorMessage={this.validateAddressField(this.props.value.zipcode) ? undefined : 'Please enter a valid ZIP code'}
+              label={this.state.requireStateZip ? 'ZIP Code' : 'Postal Code'}
```
Comment (austin-martinez) [inline src/client/components/questions/Address.jsx]: same as above, L118 also checks this condition

**Comment by akainic** [comment]
FYI just got off a call and Josh mentioned that postal code is a required field is it is a foreign country. I think we originally thought province and postal code are optional if a foreign country, but only province is optional, postal code is required.

File: src/client/components/questions/Address.jsx (hunk 1 above)
Comment (markolson) [inline src/client/components/questions/Address.jsx]: You're right - that's leftover when I was trying to figure out why validations were still failing.

**Comment by markolson** [comment]
@austin-martinez fixed the things you found

File: src/client/components/questions/Address.jsx [hunk 3]
```diff
@@ -25,12 +28,20 @@ class Address extends React.Component {
   // TODO: Look into if this is the best way to update address,
   // it is incredibly slow right now
   handleChange(path, update) {
+    // if the country is US, Canada, or Mexico, then the
+    // state/provice + postal is required
+    if (path === 'country') {
+      this.setState({ requireStateZip: _.hasIn(states, update.value) });
```
Comment (akainic) [inline src/client/components/questions/Address.jsx]: I would prefer not to store this in state if possible. Can we instead just check the value of `this.props.value.country` directly?

Comment (markolson) [inline src/client/components/questions/Address.jsx]: Sure thing

**Comment by markolson** [comment]
While testing, I found an issue with form validation I'll fix in the AM

File: src/client/utils/validations.js [hunk 4]
```diff
@@ -106,8 +106,12 @@ function isValidEmail(value) {
 // TODO:  1. what is a valid address?
 //        2. 6 arguments to a function is ugly...
 //        3. argument order is now based on form order... using
-function isValidAddress(street, city, country, state, zipcode) {
-  return isNotBlank(street) && isNotBlank(city) && isNotBlank(country) && isNotBlank(state) && isNotBlank(zipcode);
+function isValidAddress(street, city, country, state, zipcode, postalCode) {
+  const initialOk = isNotBlank(street) && isNotBlank(city) && isNotBlank(country);
+  if (country === 'USA') {
+    return initialOk && isNotBlank(state) && isNotBlank(zipcode);
```
Comment (akainic) [inline src/client/utils/validations.js]: Should we also add a validation on ZIP code here?

**Comment by markolson** [comment]
@akainic tests passed 🎉

File: src/client/utils/validations.js [hunk 5]
```diff
@@ -468,7 +475,6 @@ export {
   isValidMonetaryValue,
   isValidPhone,
   isValidEmail,
-  isValidAddress,
```
Comment (akainic) [inline src/client/utils/validations.js]: Do we need to export `isValidAddressField` here?

File: test/e2e/utils/common.js [hunk 6]
```diff
@@ -170,6 +170,12 @@ const testValues = {
   campLejeune: false
 };
 
+function selectDropdown(client, field, value) {
+  client.click(`select[name='${field}']`)
+    .click(`select option[value='${value}']`)
+    .keys(['\uE006']);
```
Comment (akainic) [inline test/e2e/utils/common.js]: Is this to fix the Selenium errors you were seeing?

File: src/client/utils/validations.js (hunk 5 above)
Comment (markolson) [inline src/client/utils/validations.js]: It's only used internally to check Vet/Spouse addresses

File: test/e2e/utils/common.js (hunk 6 above)
Comment (markolson) [inline test/e2e/utils/common.js]: Yes. Firefox has an issue with dropdown selections triggering events when being driven by selenium, and this is a bit of a workaround I found.

File: src/client/utils/validations.js (hunk 5 above)
Comment (akainic) [inline src/client/utils/validations.js]: Oh. Just realized `isValidAddressField` here isn't even being used at all? There's a different function defined in `Address` called `validateAddressField` that's actually being used.

Comment (markolson) [inline src/client/utils/validations.js]: `isValidAddressField` is called via `isValidForm`, which is called when the 'Continue' button is clicked

Comment (akainic) [inline src/client/utils/validations.js]: Ah right, thanks!

**Comment by akainic** [comment]
LGTM!

**Pull request merged**